handbook/engineering/code-intelligence/code-walkthrough.md

```diff
@@ -0,0 +1,57 @@
+# Code walkthrough
+
+The following page details how we approach the technical hiring within the Code Intelligence team instead of take home projects.
+
+### What is the day to day engineering experience at Sourcegraph?
+- We're often reading code we haven't written.
+- We're functioning in an asynchronous environment where clear communication is key.
+- We need to help our team understand areas where we are the domain expert.
+
+### What do we want to be true about this interview?
+- Allow you as the candidate to focus on an area that you are comfortable in.
+- Allow a dialogue that simulates a technical discussion between your potential future peers.
+- Use the discussion as a proxy for technical aptitude and programming skills.
+- Have a time efficient process for yourself and Sourcegraph.
+
+### We want to find engineers that are:
+- Good communicators.
+- Can tailor the technical depth and scope of a conversation dependent on the audience.
+- Can effectively teach and share their knowledge.
+
+----
+
+### Overview
+
+For this interview you'll drive a walkthrough of either a personal project or library of your choice.
+
+The following are sample ideas, the most important thing is you choose a library that you are comfortable and familiar with.  Walkthroughs can be done in
+any language, we do have a preference if possible that they are close to the domain you'd be working in.
+
+- For a full-stack position you may choose a library that you've used heavily such as [DateJs](https://github.com/datejs/Datejs)
+or to guide us through [Redux](https://github.com/reduxjs/redux).
+- If you're interviewing for a backend position you may choose a library such as [Mux](https://github.com/gorilla/mux) or [Ristretto](https://github.com/dgraph-io/ristretto).
+
+---
+- Interview length: 45 minutes
+    - 5 minutes of introductions.
+    - 30 minute walkthrough (script below).
+    - 10 minutes of questions.
+- Be ready to screenshare the library in question via your IDE or editor of choice.
+- At some point during the discussion, the engineers will ask you to describe a function of their choice in
+  which you'll walk them through the language constructs and flow.
+
+### Interview script
+
+As we're attempting to stimulate a technical discussion between peers the following list is an outline of
+how we'll approach the conversation.  The ordering below is not indicative of the flow of the discussion.
+
+- What problem does this library solve?
```

Comment:
It's good to keep this rather free-form as any script we write will not be useful for _some_ future interview we'll encounter (too many language, library, code choices to discuss).

It might be better to make a scorecard on the dimensions which we want to rate the candidate? We can still have an example script to get the format across.

Reply:
Do you have an idea for the dimensions you'd suggest?  Do you think the script should be slimmed down?  

From my initial notes these were the traits I'd listed: 

### What are the values we're testing for?

- Technical communication
	- Are they able to have a flow (describing from the outside in or vice versa)
	- Can the audience easily understand?
	- Can they teach a new concept to the team?
- Depth of understanding
	- Can they explain the code?
	- Can they explain a specific function to show language understanding?
	- Can they explain how to utilize the library/project?
	- Can they explain different levels of abstraction
	- Can they discuss tradeoffs?

Reply:
Yeah I think having an (example) question that could show each of these would be nice. What kind of question should we aim to prompt a tradeoff discussion? How should we evaluate that answer?

Reply:
> What kind of question should we aim to prompt a tradeoff discussion?
I guess there are levels to this, we could be talking lower level items or high level architectural stuff.  So I guess simply based on the area you're discussing,  _are there any tradeoffs you'd make here to achieve X_ etc.

> How should we evaluate that answer?
This is tricky and subjective, I'm not sure.  What's your take?

Reply:
I'd defer to @olafurpg and @tjdevries who went through the actual interview to help answer some of these. What kind of tradeoff questions would you find interesting with respect to metals/telescope?

My outline of an idea would be something like:

- Depth of understanding
  - Can they explain the code?
    - Give a link to a function in sg and an explanation of it at the level you'd expect from this interview (this would be super helpful if candidates have access to this document ahead of time as well so we're not constantly calibrating for depth of explanation during interview time)
  - Can they explain a specific function to show language understanding?
    - What outcome here is different than the "explain this function above" answer? How do we draw this signal out of the same answer?
  - Can they explain how to utilize the library/project?
    - Are we looking for them to write sample usage code/pseucode? What kind of follow-up questions does this afford (depending on library/project)?
  - Can they explain different levels of abstraction
    - Is there a specific question we can ask to prompt this? Should we be driving different depths of code and see if they can adjust their explanation appropriately?
  - Can they discuss tradeoffs?
    - Question seems dependent on library vs project, and whether they wrote it or not. In any case I think a question like "what if you _did alternate implementation_" and then had them list pros/cons of each at some high-ish level

Reply:
Added https://github.com/sourcegraph/about/pull/3857/commits/1621e410ccbc27b6d21060c93217950dd01b83bc per our slack discussion and your :mage: